### ISSUE TITLE
Recursive batch flush do not work

### ISSUE THREAD

**Issue by rPraml** [ebean-orm/ebean#3037]
## Expected behavior

When having a PersistController, that performs a query, I expect that it works also in batch mode

## Actual behavior

Currently, if you try to persist a larger object graph (like Customer/Contact/ContactNote), you may see

```
io.ebean.DataIntegrityException: Error when batch flush on sql: insert into contact (first_name, last_name, phone, mobile, email, is_member, cretime, updtime, customer_id, group_id) values (?,?,?,?,?,?,?,?,?,?)
NULL not allowed for column "CUSTOMER_ID";
```

This happens, if you have multiple `Customers` and `Contacts` in the batch buffer.

### Steps to reproduce

We have written a test case (and a possible workaround) here: https://github.com/FOCONIS/ebean/commit/7281602b8a306cc891ff8a6f1fb72595aafbf507

The concrete problem is, that there are
- Two Customers Customer1 and Customer2
- At least two contacts refering to Customer1 and Customer2
- Customer1 performs a batch flush.

When transaction.commit is invoked (or batch buffer is full), ebean tries to execute them in [BatchControl.executeAll](https://github.com/ebean-orm/ebean/blob/master/ebean-core/src/main/java/io/ebeaninternal/server/persist/BatchControl.java#L310)
The `bsArray` contains Customer / Contact and ContactNote. As first step, the Customer buffer is executed. This calls recursively the BatchControl.executeAll and so, an new `bsArray` is constructed with 0 customers and 2 contacts. If the recusive call happens from customer1, the second contact, that needs customer2  can not be saved.

Currently, we do not allow recursive batch flushes by disabling them in the transaction. We also tried different approaches, like checking the transaction depth. (see [here](https://github.com/FOCONIS/ebean/commit/bcd2ada3f3ebbfefad454226385a83d3ac461f27)) but this breaks other tests.

I know the current solution may be only a workaround, what we really need is some code that knows: If we are currently flushing "Cusotmers", we must not flush the dependant "Contacts" (and ContactNotes) or at maximum the contacts of the customers, that are already saved.
@rbygrave maybe you have an other idea, how this can be solved. For me, the soluition with disabling `flushOnQuery` should work, but I think, this is not a good solution.

Roland

**Comment by rbygrave** [comment]
Is this fixed by https://github.com/ebean-orm/ebean/pull/3148 ?

**Comment by rPraml** [comment]
Yes, fixed by #3148

**Comment by deverlex** [comment]
javax.persistence.PersistenceException: Error when batch flush on sql: insert into order_items (...

Caused by: java.sql.BatchUpdateException: (conn=18) Column 'order_id' cannot be null

	at org.mariadb.jdbc.ClientPreparedStatement.executeBatchPipeline(ClientPreparedStatement.java:196)
	at org.mariadb.jdbc.ClientPreparedStatement.executeInternalPreparedBatch(ClientPreparedStatement.java:128)
	at org.mariadb.jdbc.ClientPreparedStatement.executeBatch(ClientPreparedStatement.java:466)
	at com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:127)

It's not fixed at version 13.25.1
